http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/ProxyResolver.scala

```diff
@@ -0,0 +1,54 @@
+package org.broadinstitute.dsde.workbench.leonardo.dns
+
+import akka.http.scaladsl.model.Uri.Host
+import cats.effect.Effect
+import cats.effect.concurrent.Ref
+import cats.effect.implicits._
+import cats.implicits._
+import org.broadinstitute.dsde.workbench.model.IP
+import org.http4s.Uri
+import org.http4s.client.RequestKey
+
+import java.net.InetSocketAddress
+import scala.concurrent.Future
+
+/**
+ * Implements custom DNS resolution for accessing routes through the Leo proxy.
+ * Contains APIs for http4s and akka-http.
+ */
+trait ProxyResolver[F[_]] {
+  // http4s API
+  // See https://github.com/http4s/http4s/pull/4699
+  def resolveHttp4s(requestKey: RequestKey): Either[Throwable, InetSocketAddress]
+
+  // akka-http API
+  // See https://doc.akka.io/docs/akka-http/current/client-side/client-transport.html#custom-host-name-resolution-transport
+  def resolveAkka(host: String, port: Int): Future[InetSocketAddress]
+}
+
+/**
+ * Implementation of ProxyResolver using a Map[Host, IP] stored in a Ref.
+ */
+class ProxyResolverInterp[F[_]](hostToIpMapping: Ref[F, Map[Host, IP]])(
```

Comment:
Make this private?(just to make sure this is not being used as arguments in any class definition directly in other places)  and make a constructor for this in `ProxyResolver`'s companion object?

Reply:
Sure 👍